org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/util/CharDeduplication.java

```diff
@@ -35,31 +34,17 @@ public class CharDeduplication {
 	/** number of entries to linear search affects performance but decreases collisions - does not affect memory */
 	public static final int SEARCH_SIZE = 8; // a power of 2, has to be smaller then TABLE_SIZE
 
-	/** avoid OOME by additional CharDeduplication memory **/
-	static final class CacheReference<T> {
-		private SoftReference<T> reference;
-		private final Supplier<? extends T> supplier;
-
-		CacheReference(Supplier<? extends T> supplier) {
-			this.supplier = supplier;
-			this.reference = new SoftReference<>(supplier.get());
-		}
-
-		T get() {
-			T referent = this.reference.get();
-			if (referent == null) {
-				referent = this.supplier.get();
-				this.reference = new SoftReference<>(referent);
-			}
-			return referent;
-		}
-	}
-
-	private final static ThreadLocal<CacheReference<CharDeduplication>> mutableCache = ThreadLocal.withInitial(()->new CacheReference<>(CharDeduplication::new));
+	private final static ThreadLocal<SoftReference<CharDeduplication>> mutableCache = ThreadLocal
+			.withInitial(() -> new SoftReference<>(new CharDeduplication()));
 
 	/** @return an instance that is *not* thread safe. To be used in a single thread only. **/
 	public static CharDeduplication getThreadLocalInstance() {
-		return mutableCache.get().get();
+		CharDeduplication local = mutableCache.get().get();
```

Comment:
this looks not atomic. i got the feeling it does not need to be atomic because of threadlocal. but the charm of the former implementation was that i did not have to think about it.

Reply:
What do you mean by atomic? Atomic means that an action either completes successful or not happen at all including all side effects. the former implementation did exactly the same thing.

If you mean threadsafe instead, then yes this is threadsafe because it uses a Threadlocal (that previously was just wrapped in CacheReference ...